add v1 release blog (#36)
localstack/localstack.github.io: content/blog/2022-07-12-announcing-localstack-10-general-availability/index.md

---
title: Announcing LocalStack 1.0 General Availability!
description: Announcing LocalStack 1.0 General Availability!
lead: Announcing LocalStack 1.0 General Availability!
date: 2022-07-12T11:00:55+05:30
lastmod: 2022-07-12T11:00:55+05:30
images: []
contributors: ["LocalStack Team"]
tags: ['news']
---

Today, we are excited to announce the general availability of LocalStack 1.0. This major release is a significant milestone towards our vision to propel developer productivity - by allowing dev teams to quickly and conveniently develop & test their cloud applications locally and across the CI/CD pipeline.

> LocalStack 1.0 is available with our Docker image, shipped for both our Community and Pro users. LocalStack's core emulation platform provides emulated cloud APIs, currently primarily focused on the AWS cloud. It is presently being used by a large and active open source community with over 100K active users worldwide. With various features for individual productivity and team collaboration, we provide a comprehensive ecosystem of local AWS services, integrations and tools to make cloud development a breeze.

Reviewer: ```suggestion
LocalStack 1.0 is available with our Docker image, shipped for both our Community and Pro users, as well as all users of our newly introduced Team product tier. LocalStack's core emulation platform provides emulated cloud APIs, currently primarily focused on AWS cloud. It is presently being used by a large and active open source community with over 100K active users worldwide. With various features for individual productivity and team collaboration, we provide a comprehensive ecosystem of local AWS services, integrations and tools to make cloud development a breeze.
```

> With LocalStack 1.0, we mark the first milestone of LocalStack's mission to become the go-to platform for local cloud development. We have spent the last year significantly re-shaping the codebase to make it easier to introduce and extend AWS services, improving parity with AWS, introducing mechanisms to monitor parity, and adding new Pro features. This blog looks at what's new in LocalStack and what it means for our community and users.

Reviewer: ```suggestion
With LocalStack 1.0, we mark the first milestone of LocalStack's mission to become the go-to platform for local cloud development. We have spent the last year significantly re-shaping the codebase to make it easier to introduce and extend AWS services, improving parity with AWS, introducing mechanisms to monitor parity, adding new Pro features and introducing a completely new Team tier focused on cross team collaboration. This blog looks at what's new in LocalStack and what it means for our community and users.
```

## What’s new in LocalStack?

> LocalStack started as an open-source project in August 2016, with the initial commit adding support for eight-core AWS APIs, including API Gateway, Lambda, DynamoDB, and a few others. With the v1 release, we now support over 80 services distributed across our community and pro versions to power your local development and testing productivity and better cover your use-cases ranging from individual needs to enterprise use-cases.  

Reviewer: ```suggestion
LocalStack started as an open-source project in August 2016, with the initial commit adding support for eight-core AWS APIs, including API Gateway, Lambda, DynamoDB, and a few others. With the v1 release, we now support over 80 services distributed across our community and pro versions as well as a number of advanced team collaboration features in our new product tier to power your local development and testing productivity and better cover your use-cases ranging from individual needs to wide enterprise use-cases.  
```

A 1.0 release for LocalStack is a significant milestone for us after over 1,000,000 PyPI downloads and over 100,000,000 Docker pulls. LocalStack now satisfied the requirements we laid out in our previous blog to fix the broken cloud software development model. LocalStack currently operates with greater interoperability with the real AWS cloud, provides stability to our APIs and protocols, introduces new ways of reporting parity with AWS, persistence support and much more!

These features include:

> ### New Pro Features

Reviewer: ```suggestion
### New Pro and Team Features
```

Use cases for extensions include:

-   Starting custom services together with LocalStack in the same container (see our [LocalStack Stripe example](https://github.com/localstack/localstack-extensions/tree/main/stripe))
-   Instrumenting AWS requests with additional information before they reach your Lambdas (see [case study with Thundra](https://localstack.cloud/blog/2021-09-16-test-monitoring-for-localstack-apps/))
-   Logging AWS API calls to custom data backends
-   … and much more!

### IAM enforcement for all services

LocalStack now supports IAM enforcement for all supported AWS services! In addition to that, you can expect more parity regarding the required actions for requests, and our Explainable IAM feature will tell you which declarations are missing from your policies! IAM enforcement can be activated with ENFORCE_IAM=1. You can find more information in our [IAM documentation](https://docs.localstack.cloud/aws/iam/).

### Detailed stack analytics

> As part of an upcoming Pro feature we call “Stack Insights”, we have instrumented LocalStack to report AWS API usage telemetry of LocalStack runs to your LocalStack Pro account. We can now show you which APIs you are using, which clients of your integrations use particular services and API operations, and which services cause the most API errors, and more. No sensitive data of your stacks are collected! Over the next weeks we will be rolling out new UI features to give you fine-grained access to this data.

Reviewer: ```suggestion
As part of an upcoming Team feature we call “Stack Insights”, we have instrumented LocalStack to report AWS API usage telemetry of LocalStack runs to your LocalStack Team account. We can now show you which APIs you are using, which clients of your integrations use particular services and API operations, and which services cause the most API errors, and more. No sensitive data of your stacks are collected! Over the next weeks we will be rolling out new UI features to give you fine-grained access to this data.
```

AWS requests are now logged uniformly in the `INFO` log level (set by default or when `DEBUG=0`). The shape is `AWS <service>.<operation> => <http-status> (<error type>)`. Requests to HTTP endpoints are logged in a similar way.

```shell
2022-07-12T10:12:03.250  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS s3.PutObject => 404 (NoSuchBucket)
2022-07-12T10:12:11.295  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS s3.CreateBucket => 200
2022-07-12T10:12:13.159  INFO --- [   asgi_gw_1] localstack.request.aws     : AWS s3.PutObject => 200
2022-07-12T10:12:28.761  INFO --- [   asgi_gw_0] localstack.request.http    : GET /_localstack/health => 200
```

If you want access to detailed log output, including the input and output of requests, you can start LocalStack with `LS_LOG=trace`, which will provide the same log format, but append the request and response objects, as well as the HTTP headers to the log line.

When `DEBUG=1` is enabled, errors inside LocalStack are now reported to the client in full. When writing issue reports, these stack traces help LocalStack contributors to better triage your issues.

### Improved persistence support

> We are introducing an enhanced persistence mechanism that allows you to preserve state across LocalStack container restarts (available in the Pro version).

Reviewer: ```suggestion
We are introducing an enhanced persistence mechanism that allows you to preserve your state across LocalStack container restarts (available in the Pro version).
```

> In the past, there used to be a naive persistence implementation based on record&replay of API calls. This mechanism had several conceptual flaws and was hence deprecated in version. The all-new persistence experience is much more performand and more reliable.

Reviewer: ```suggestion
In the past, there used to be a naive persistence implementation based on record&replay of API calls. This mechanism had several conceptual flaws and was hence deprecated in version. The all-new persistence experience is much more performant and more reliable.
```
